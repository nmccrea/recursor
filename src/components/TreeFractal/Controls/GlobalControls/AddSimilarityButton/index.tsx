import React from "react"
import { useDispatch } from "react-redux"
import { addOne } from "../../../../../state/similarities/actions"
import { SimilarityInputs } from "../../../../../state/similarities/types"

/**
 * Generate a set of new similarity inputs.
 *
 * @returns A new `SimilarityInputs` object that may be added to the store.
 */
const generateNewSimilarityInputs = (): SimilarityInputs => ({
  translation: 1,
  scale: 0.75,
  angle: 0,
  depth: 5,
  color: "black",
})

const AddSimilarityButton = () => {
  const dispatch = useDispatch()
  return (
    <button onClick={() => dispatch(addOne(generateNewSimilarityInputs()))}>
      Add Similarity
    </button>
  )
}

export default AddSimilarityButton
