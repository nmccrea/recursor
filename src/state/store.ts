/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { configureStore, combineReducers } from "@reduxjs/toolkit"
import similarities from "./similarities/reducer"

const rootReducer = combineReducers({ similarities })
const createStore = () => configureStore({ reducer: rootReducer })

export { createStore }
export type RootState = ReturnType<typeof rootReducer>
export type Selector<ReturnType> = (state: RootState) => ReturnType | undefined
