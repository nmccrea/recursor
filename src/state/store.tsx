/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { combineReducers, configureStore } from "@reduxjs/toolkit"

const rootReducer = combineReducers({})
export type RootState = ReturnType<typeof rootReducer>
export default configureStore({ reducer: rootReducer })
