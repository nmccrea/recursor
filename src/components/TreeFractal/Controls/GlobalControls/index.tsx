import React from "react"
import AddSimilarityButton from "./AddSimilarityButton"

const GlobalControls = () => (
  <div>
    <AddSimilarityButton />
  </div>
)

export default GlobalControls
