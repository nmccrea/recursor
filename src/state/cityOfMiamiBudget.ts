/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import { RootState } from "./store"
import {
  createAsyncThunk,
  createSlice,
  CaseReducer,
  PayloadAction,
} from "@reduxjs/toolkit"

// State
enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}

interface CityOfMiamiBudgetState {
  status: Status
  data?: object
}

const INITIAL_STATE: CityOfMiamiBudgetState = { status: Status.Idle }

// Case Reducers
const fetchDataPending: CaseReducer<CityOfMiamiBudgetState, PayloadAction> = (
  previousState
) => ({
  ...previousState,
  status: Status.Pending,
})

const fetchDataFulfilled: CaseReducer<
  CityOfMiamiBudgetState,
  PayloadAction<object>
> = (previousState, action) => ({
  ...previousState,
  status: Status.Fulfilled,
  data: action.payload,
})

// Async Thunks
const fetchData = createAsyncThunk("fetchData", async (url: string) => {
  const response = await fetch(url)
  return await response.json()
})

// Slice
const { reducer } = createSlice({
  name: "cityOfMiamiBudget",
  initialState: INITIAL_STATE,
  reducers: {},
  extraReducers: {
    [fetchData.pending.type]: fetchDataPending,
    [fetchData.fulfilled.type]: fetchDataFulfilled,
  },
})

// Selectors
const getStatus = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.status) || undefined

const getData = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.data) || undefined

export default reducer
export { fetchData, getStatus, getData, Status }
