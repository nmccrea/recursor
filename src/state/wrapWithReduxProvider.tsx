/**
 * Implements Gatsby's `wrapRootElement` API.
 * @packageDocumentation
 */

import React from "react"
import { Provider as ReduxProvider } from "react-redux"
import store from "./store"

interface Arg {
  element: React.ReactNode
}
export const wrapWithReduxProvider = ({ element }: Arg) => (
  <ReduxProvider store={store}>{element}</ReduxProvider>
)
