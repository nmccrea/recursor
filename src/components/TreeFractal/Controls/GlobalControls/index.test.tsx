import React from "react"
import { render } from "@testing-library/react"
import { Provider } from "react-redux"
import { createStore } from "../../../../state/store"
import GlobalControls from "."

describe("`<GlobalControls />`", () => {
  it("renders correctly", () => {
    const { container } = render(
      <Provider store={createStore()}>
        <GlobalControls />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })
})
