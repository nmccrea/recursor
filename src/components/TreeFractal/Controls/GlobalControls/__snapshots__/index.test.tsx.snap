// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<GlobalControls />\` renders correctly 1`] = `
<div>
  <div>
    <button
      class="sc-AxjAm sc-AxirZ CUWHb"
    >
      Add
    </button>
  </div>
</div>
`;
